From fa274b0ea78fb28d3b8f98e1d81c98afd5a2fd11 Mon Sep 17 00:00:00 2001
From: Zi Shen Lim <zlim@netlogicmicro.com>
Date: Mon, 3 Oct 2011 11:20:51 -0700
Subject: [PATCH 432/762] Fix compile error: mtd

mtd_device_{,un}register replaced {add,del}_mtd_partitions in mainline.
mtd_has_partitions check deprecated in mainline.

existing code doesn't del_mtd_partitions - revisit
also to revisit logic for xlp_plat_nand_probe.

Based on Broadcom SDK 2.3.

Signed-off-by: Zi Shen Lim <zlim@netlogicmicro.com>
Signed-off-by: Yanjiang Jin <yanjiang.jin@windriver.com>
---
 drivers/mtd/nand/xlp_plat_nand.c |   33 +++++++++++++++------------------
 1 files changed, 15 insertions(+), 18 deletions(-)

diff --git a/drivers/mtd/nand/xlp_plat_nand.c b/drivers/mtd/nand/xlp_plat_nand.c
index 05d55b9..7089287 100644
--- a/drivers/mtd/nand/xlp_plat_nand.c
+++ b/drivers/mtd/nand/xlp_plat_nand.c
@@ -726,26 +726,22 @@ static int __devinit xlp_plat_nand_probe(struct platform_device *pdev)
 #endif
 
 #ifdef CONFIG_MTD_PARTITIONS
-	if (mtd_has_partitions()) {
-                if (mtd_has_cmdlinepart()) {
-                        static const char *part_probes[] = { "cmdlinepart", NULL, };
-                        data->nr_parts = parse_mtd_partitions(&data->mtd, part_probes, &data->parts, 0);
-                }
-                if(data->nr_parts <= 0)
-                {
-                        data->nr_parts = ARRAY_SIZE(xlp_nand_partition_info);
-                        if(!data->parts)
-                                data->parts = xlp_nand_partition_info;
-                }
-                if(data->nr_parts > 0)
-                {
-			/* Register the partitions */
-                        res = add_mtd_partitions(&data->mtd, data->parts, data->nr_parts);
-                }
-        }
+	if (mtd_has_cmdlinepart()) {
+		static const char *part_probes[] = { "cmdlinepart", NULL, };
+		data->nr_parts = parse_mtd_partitions(&data->mtd, part_probes, &data->parts, 0);
+	}
+	if (data->nr_parts <= 0) {
+		data->nr_parts = ARRAY_SIZE(xlp_nand_partition_info);
+		if (!data->parts)
+			data->parts = xlp_nand_partition_info;
+	}
+	if (data->nr_parts > 0) {
+		/* Register the partitions */
+		res = mtd_device_register(&data->mtd, data->parts, data->nr_parts);
+	}
 	else
 #endif
-		res = add_mtd_device(&data->mtd);
+		res = mtd_device_register(&data->mtd, NULL, 0);
 
         if (!res)
                 return res;
@@ -768,6 +764,7 @@ static int __devexit xlp_plat_nand_remove(struct platform_device *pdev)
 
         nand_release(&data->mtd);
 #ifdef CONFIG_MTD_PARTITIONS
+	/* XLP_MERGE_TODO: mtd_device_unregister? */
         if (data->parts && data->parts != pdata->chip.partitions)
                 kfree(data->parts);
 #endif
-- 
1.7.0.4

